include/deal.II/lac/solver_cg.h

```diff
@@ -389,6 +462,481 @@ SolverCG<VectorType>::compute_eigs_and_cond(
+      {
+        // Initialize without setting the vector entries, as those would soon
+        // be overwritten anyway
+        r.reinit(x, true);
+        p.reinit(x, true);
+        v.reinit(x, true);
+        if (flexible)
+          z.reinit(x, true);
+
+        // compute residual. if vector is zero, then short-circuit the full
+        // computation
+        if (!x.all_zero())
+          {
+            A.vmult(r, x);
+            r.sadd(-1., 1., b);
+          }
+        else
+          r.equ(1., b);
+
+        residual_norm = r.l2_norm();
+      }
+
+      void
+      do_iteration(const unsigned int iteration_index)
+      {
+        const Number previous_r_dot_preconditioner_dot_r =
+          r_dot_preconditioner_dot_r;
+        previous_alpha = alpha;
+        previous_beta  = beta;
+
+        if (std::is_same<PreconditionerType, PreconditionIdentity>::value ==
+            false)
+          {
+            preconditioner.vmult(v, r);
+            r_dot_preconditioner_dot_r = r * v;
+          }
+        else
+          r_dot_preconditioner_dot_r = residual_norm * residual_norm;
+
+        const VectorType &direction =
+          std::is_same<PreconditionerType, PreconditionIdentity>::value ? r : v;
+
+        if (iteration_index > 1)
+          {
+            Assert(std::abs(previous_r_dot_preconditioner_dot_r) != 0.,
+                   ExcDivideByZero());
+            beta =
+              r_dot_preconditioner_dot_r / previous_r_dot_preconditioner_dot_r;
+            if (flexible)
+              beta -= (r * z) / previous_r_dot_preconditioner_dot_r;
+            p.sadd(beta, 1., direction);
+          }
+        else
+          p.equ(1., direction);
+
+        if (flexible)
+          z.swap(v);
+
+        A.vmult(v, p);
+
+        const Number p_dot_A_dot_p = p * v;
+        Assert(std::abs(p_dot_A_dot_p) != 0., ExcDivideByZero());
+        alpha = r_dot_preconditioner_dot_r / p_dot_A_dot_p;
+
+        x.add(alpha, p);
+        residual_norm = std::sqrt(std::abs(r.add_and_dot(-alpha, v, r)));
+      }
+
+      void
+      finalize_after_convergence(const unsigned int)
+      {}
+    };
+
+
+
+    // Actual class with the basic operation implemented in
+    // IterationWorkerBase
+    template <typename VectorType,
+              typename MatrixType,
+              typename PreconditionerType,
+              typename = int>
+    struct IterationWorker
+      : public IterationWorkerBase<VectorType, MatrixType, PreconditionerType>
+    {
+      IterationWorker(const MatrixType &        A,
+                      const PreconditionerType &preconditioner,
+                      const bool                flexible,
+                      VectorMemory<VectorType> &memory,
+                      VectorType &              x)
+        : IterationWorkerBase<VectorType, MatrixType, PreconditionerType>(
+            A,
+            preconditioner,
+            flexible,
+            memory,
+            x)
+      {}
```

Comment:
That's a good idea, I move everything that belongs to the generic case here. The idea of the base class was to collect common code and variable in a single place, to avoid diverging interfaces.